UIKit/Gradient/TWTGradient.m

```diff
@@ -0,0 +1,205 @@
+}
+
+
+- (void)dealloc
+{
+    if (_CGGradient) {
+        CFRelease(_CGGradient);
+    }
+}
+
+
+- (NSString *)description
+{
+    return [NSString stringWithFormat:@"<%@ %p: gradientColors=%@>", self.class, self, self.gradientColors];
+}
+
+
+- (NSUInteger)hash
+{
+    return self.gradientColors.hash;
+}
+
+
+- (BOOL)isEqual:(id)object
+{
+    typeof(self) other = object;
+    return [object isKindOfClass:self.class] && [other.gradientColors isEqualToArray:self.gradientColors];
+}
+
+
+#pragma mark - NSCopying
+
+- (id)copyWithZone:(NSZone *)zone
+{
+    return self;
+}
+
+
+#pragma mark - NSSecureCoding
+
++ (BOOL)supportsSecureCoding
+{
+    return YES;
+}
+
+
+- (instancetype)initWithCoder:(NSCoder *)decoder
+{
+    NSArray<TWTGradientColor *> *gradientColors = [decoder decodeObjectOfClass:[NSArray class] forKey:NSStringFromSelector(@selector(gradientColors))];
+    return [self initWithGradientColors:gradientColors];
+}
+
+
+- (void)encodeWithCoder:(NSCoder *)encoder
+{
+    [encoder encodeObject:self.gradientColors forKey:NSStringFromSelector(@selector(gradientColors))];
+}
+
+
+#pragma mark - Gradient Image Generation
+
+- (UIImage *)drawImageWithSize:(CGSize)size block:(void (^ _Nonnull)(CGContextRef _Nonnull))block
+{
+    UIGraphicsBeginImageContextWithOptions(size, NO, 0);
+    CGContextRef context = UIGraphicsGetCurrentContext();
+    block(context);
+    UIImage *image = UIGraphicsGetImageFromCurrentImageContext();
+    UIGraphicsEndImageContext();
+    return image;
+}
+
+
+- (UIImage *)linearGradientImageWithSize:(CGSize)size startPoint:(CGPoint)startPoint endPoint:(CGPoint)endPoint
+{
+    return [self drawImageWithSize:size block:^(CGContextRef _Nonnull context) {
+        CGContextDrawLinearGradient(context, self.CGGradient, startPoint, endPoint, kCGGradientDrawsBeforeStartLocation | kCGGradientDrawsAfterEndLocation);
+    }];
+}
+
+
+- (UIImage *)radialGradientImageWithSize:(CGSize)size center:(CGPoint)center radius:(CGFloat)radius
+{
+    return [self drawImageWithSize:size block:^(CGContextRef _Nonnull context) {
+        CGContextDrawRadialGradient(context, self.CGGradient, center, 0, center, radius, kCGGradientDrawsBeforeStartLocation | kCGGradientDrawsAfterEndLocation);
+    }];
+}
+
+@end
+
+
+#pragma mark - Gradient Color
+
+@implementation TWTGradientColor
+
+- (instancetype)initWithColor:(UIColor *)color location:(CGFloat)location
+{
+    NSParameterAssert(color);
+
+    self = [super init];
+    if (self) {
+        _color = color;
+        _location = MIN(MAX(0.0, location), 1.0);
+    }
+
+    return self;
+}
+
+
+- (instancetype)initWithHexadecimalRGB:(NSUInteger)hex location:(CGFloat)location
+{
+    CGFloat red = (CGFloat)((hex & 0xff0000) >> 16) / 255.0;
+    CGFloat green = (CGFloat)((hex & 0x00ff00) >> 8) / 255.0;
+    CGFloat blue = (CGFloat)(hex & 0x0000ff) / 255.0;
+    UIColor *color = [UIColor colorWithRed:red green:green blue:blue alpha:1.0];
+    return [self initWithColor:color location:location];
+}
```

Comment:
This seems out of place. We have the color category already. Can we get rid of this?